Create rule S5147[Java]: NoSQL operations should not be vulnerable to injections (#668)
SonarSource/rspec: rules/S5147/description.adoc

User-supplied data such as URL parameters and the content of POST should always
be considered untrustworthy and tainted.

> Applications that perform NoSQL operations based on unvalidated tainted data
> can be exploited similarly to regular SQL injection bugs.

Reviewer: Unvalidated = tainted 
```suggestion
Applications that perform NoSQL operations based on tainted data can be exploited similarly to regular SQL injection bugs.
```

Depending on the code, the same risks apply: The attacker aims to access
sensitive information or compromise data integrity. To achieve this, the
attacks may involve the misuse of query operators, JavaScript code, or simple

> string operations.

Reviewer: Sorry but I think it's longer and less clear that "an attacker could inject NoSQL objects to access sensitive information or compromise data integrity."

Author: My goal here is to separate different information, namely the risks (the why of the attacker), and the how(its technics). The initial sentence is really long and could use a little breath between informations

```suggestion
Depending on the code, the same risks exist as with SQL injections: The attacker
aims to access sensitive information or compromise data integrity. The
attacks may involve the injection of query operators, JavaScript code, or
string operations.
```

string operations.

> The problem could be mitigated by ensuring the type of the input is a String or sanitizing the user-provided data.

Reviewer: "ensuring the type of the input is a String" for JavaScript I think it still make sens but nit for Java.

> This problem can be mitigated by using an Object Document Mapper (ODM) library

Reviewer: Which one? Maybe MongoDB Java Driver have an ODM API?
If not we should advise an ODM library and maybe add a code example to illustrate it.
Maybe a more practical advise would be to only use the "safe" API from MongoDB Java Driver

Author: https://github.com/MorphiaOrg/morphia

Author: I added them in the See section, I don't think it makes sense to add them in the code snippets though, since we add already a "native" example

> or by validating user-supplied data based on its size or allowable characters.

Reviewer: "allowable characters" sound strange. Why not "allowed characters"?

Author: ```suggestion
or by validating user-supplied data based on its size or allowed characters.
```

User-provided data such as URL parameters and POST body-content should always
be considered untrusted and tainted.

Applications that perform NoSQL operations based on tainted data can be
exploited similarly to regular SQL injection bugs. Depending on the code, the
same risks exist as with SQL injections: The attacker aims to access sensitive
information or compromise data integrity. Attacks may involve the injection of

> query operators, JavaScript code, or string operations.

Reviewer: Please remove necessary new lines

This problem can be mitigated by using an Object Document Mapper (ODM) library
or by validating user-supplied data based on its size or allowed characters.
In languages where different types can be inserted into HTTP parameters (e.g.,

> Javascript and PHP), user-supplied data should be validated as strings.

Reviewer: Language specific content should be migrated to a language specific `adoc` file

Author: Does it matter for such a short line ?

Author: done
